identity-iota/src/tangle/resolver.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+#[cfg(not(feature = "arc"))]
+use std::rc::Rc;
+#[cfg(feature = "arc")]
+use std::sync::Arc;
+
+use dashmap::DashMap;
+
+use identity_core::common::Url;
+use identity_credential::credential::Credential;
+use identity_credential::presentation::Presentation;
+
+use crate::chain::ChainHistory;
+use crate::chain::DocumentHistory;
+use crate::did::IotaDID;
+use crate::diff::DiffMessage;
+use crate::document::ResolvedIotaDocument;
+use crate::error::Error;
+use crate::error::Result;
+use crate::tangle::Client;
+use crate::tangle::ClientBuilder;
+use crate::tangle::NetworkName;
+use crate::tangle::TangleResolve;
+
+#[cfg(not(feature = "arc"))]
+type ClientRc = Rc<Client>;
+
+#[cfg(feature = "arc")]
+type ClientRc = Arc<Client>;
+
+/// A `Resolver` supports resolving DID Documents across different Tangle networks using
+/// multiple [`Clients`][Client].
+///
+/// Also provides convenience functions for resolving DID Documents associated with
+/// verifiable [`Credentials`][Credential] and [`Presentations`][Presentation].
+#[derive(Debug)]
+pub struct Resolver {
+  // TODO: is DashMap really necessary here?
+  //  We can avoid mutation if we disallow adding/removing Clients post-creation.
```

Comment:
Makes a lot of sense to me to setup a resolver once at build-time and then disallow mutation afterwards. We don't have to commit to allowing mutability now, but if the need arises later, adding it is non-breaking.

Reply:
Cool, I guess we need the `ResolverBuilder` in the Wasm bindings then? It was not ported since there is no good way to avoid the builder pattern in Wasm in this case, but it would be the only option if we remove the mutators.

Reply:
Can we not take a list of `ClientOrBuilder`s in `WasmResolver::new`, or even just `Client`s? Or is that the not-good-way you meant :smile:?

```ts
const resolver = new Resolver([
      Client.fromConfig(Config.fromNetwork(Network.mainnet())),
      Client.fromConfig(Config.fromNetwork(Network.devnet())),
  ]);
```

Is the problem that someone could provide two clients with the same network name?

Reply:
As far as I know there is no way to take `Vec<WasmClient>` as a parameter (`Vec<T>` is not supported in general), and if I recall correctly we cannot cast `JsValue` to one of our exported structs, and since `Client` and `ClientBuilder` are not serializable (for good reason), we cannot use a duck-typed Typescript interface either.

I'll try again just to be sure though.

Reply:
Ah, yeah, my bad, I was thinking along the lines of the `serde` solution, but of course that doesn't work with `Client`s. Perhaps a map would work alternatively, where the keys are network names and the values `Client`s. Using [`Reflect::own_keys`](https://docs.rs/js-sys/latest/js_sys/Reflect/fn.own_keys.html) we could iterate the keys and retrieve the clients. That would even prevent duplication.

Reply:
I think that returns the keys of the map/object, not the values (`Clients`)? 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect/ownKeys#using_reflect.ownkeys

Reply:
Yes, this would actually be pretty ugly typing-wise, as we'd have to follow-up with `Reflect::get` which only returns a `JsValue` as you pointed out, and we'd have to downcast that to `Client` similar to how WasmStorage does that. That should be avoided if possible, though. Exposing the builder seems like the best option after all. Sorry for the half-baked ideas.

Reply:
Ideally all this would be avoided if `wasm-bindgen` supported `dyn_ref`/`dyn_into` on exported types, but it will probably never happen: https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/trait.JsCast.html#method.dyn_into

Wow that downcasting workaround is scary. I mean it works but there's got to be a better way than relying on the constructor name: https://github.com/rustwasm/wasm-bindgen/issues/2231#issuecomment-656293288

Reply:
Exposed `WasmResolverBuilder`. Using it looks like this in Javascript:

```ts
const client = new Client();

const resolver = await Resolver
        .builder()
        .client(client)
        .clientConfig(Config.fromNetwork(Network.devnet()))
        .build();
```